### PR TITLE
Forbid unused property renaming in destructuring binding in function types

### LINKED ISSUE #37454: didn't have wran message when use bad syntax
**TypeScript Version:**  3.7.x-dev.201xxxxx

**Search Terms:**

**Code**

```ts
export interface IA
{
	({
		a: string // <=== bad syntax
	}): any
}
```

**Expected behavior:**

show error message

**Actual behavior:**

nope

**Playground Link:**  

**Related Issues:**

### PR DESCRIPTION
**Pull request by uhyo**
Happy hacktoberfest!! 🎃🥳

Fixes #37454.

The PR implements mainly two things, as suggested in https://github.com/microsoft/TypeScript/pull/41044#issuecomment-1146271768:

- Emit an error for property renaming in destructing binding in function types if it is unused. This is not under `noUnusedLocals` nor `noUnusedParameters`, but declaration files are not checked.
- Changes declaration emit so that property renaming in destructing bunding is removed if it is unused.

## Example

```ts
type O = { a?: string; b: number; c: number; };
type F1 = (arg: number) => any; // OK
type F2 = ({ a: string }: O) => any; // Error
type F3 = ({ a: string, b, c }: O) => any; // Error
type F4 = ({ a: string }: O) => any; // Error
type F5 = ({ a: string, b, c }: O) => any; // Error
type F6 = ({ a: string }) => typeof string; // OK
type F7 = ({ a: string, b: number }) => typeof number; // Error
type F8 = ({ a, b: number }) => typeof number; // OK
type F9 = ([a, b, c]) => void; // Error

type G1 = (arg: number) => any; // OK
type G2 = ({ a: string }: O) => any; // Error
type G3 = ({ a: string, b, c }: O) => any; // Error
type G4 = ({ a: string }: O) => any; // Error
type G5 = ({ a: string, b, c }: O) => any; // Error
type G6 = ({ a: string }) => typeof string; // OK
type G7 = ({ a: string, b: number }) => typeof number; // Error
type G8 = ({ a, b: number }) => typeof number; // OK
type G9 = ([a, b, c]) => void; // Error
```

## Results

All the errors below are new in this PR.

```
src/index.ts:3:17 - error TS2842: Variable 'string' is not used. Did you mean to write a type annotation for whole object?

3 type F2 = ({ a: string }: O) => any; // Error
                  ~~~~~~

src/index.ts:4:17 - error TS2842: Variable 'string' is not used. Did you mean to write a type annotation for whole object?

4 type F3 = ({ a: string, b, c }: O) => any; // Error
                  ~~~~~~

src/index.ts:5:17 - error TS2842: Variable 'string' is not used. Did you mean to write a type annotation for whole object?

5 type F4 = ({ a: string }: O) => any; // Error
                  ~~~~~~

src/index.ts:6:17 - error TS2842: Variable 'string' is not used. Did you mean to write a type annotation for whole object?

6 type F5 = ({ a: string, b, c }: O) => any; // Error
                  ~~~~~~

src/index.ts:8:17 - error TS2842: Variable 'string' is not used. Did you mean to write a type annotation for whole object?

8 type F7 = ({ a: string, b: number }) => typeof number; // Error
                  ~~~~~~

src/index.ts:13:17 - error TS2842: Variable 'string' is not used. Did you mean to write a type annotation for whole object?

13 type G2 = ({ a: string }: O) => any; // Error
                   ~~~~~~

src/index.ts:14:17 - error TS2842: Variable 'string' is not used. Did you mean to write a type annotation for whole object?

14 type G3 = ({ a: string, b, c }: O) => any; // Error
                   ~~~~~~

src/index.ts:15:17 - error TS2842: Variable 'string' is not used. Did you mean to write a type annotation for whole object?

15 type G4 = ({ a: string }: O) => any; // Error
                   ~~~~~~

src/index.ts:16:17 - error TS2842: Variable 'string' is not used. Did you mean to write a type annotation for whole object?

16 type G5 = ({ a: string, b, c }: O) => any; // Error
                   ~~~~~~

src/index.ts:18:17 - error TS2842: Variable 'string' is not used. Did you mean to write a type annotation for whole object?

18 type G7 = ({ a: string, b: number }) => typeof number; // Error
                   ~~~~~~


Found 10 errors in the same file, starting at: src/index.ts:3
```

### CODE REVIEW

**Comment by sandersn** [comment]
@rbuckton this may be affected when you change the way binding patterns are handled.

**Comment by andrewbranch** [comment]
> This PR also modifies declaration emission so that it does not generate invalid function types as to the above rule.

Note that this PR is a breaking change, so can’t be merged into 4.1—it means people consuming existing declarations generated from totally reasonable code (your `const k = ({x: z = 'y'}) => { }` example) will suddenly get errors. We might even want to consider suppressing this error in declaration files, as it’s pretty annoying to start issuing a bunch of new errors inside people’s node_modules, with obviously no benefit to the owners of those node_modules.

**Comment by andrewbranch** [comment]
Perhaps a gentler way of doing this is putting it under `--noUnusedLocals`, since that’s basically what it is—a new name introduced that is highly unlikely to be used (though it technically can be: `type F = ({ a: foo }) => typeof foo`). `--noUnusedLocals` checks are downgraded to suggestion diagnostics in declaration files, so it has the advantage of being visible in the editor for someone who’s hand-authoring a declaration file, but will never break a build for declaration consumers.

**Comment by andrewbranch** [comment]
Wow, this PR really came back from the dead! 🧟

**Comment by andrewbranch** [comment]
Still technically a breaking change, so let’s look in the 4.6 timeline. Thanks!

**Comment by sandersn** [comment]
@andrewbranch your last comment makes it sound like this is ready to be merged, but early in a release cycle. Is that true? Am I misreading it?

**Comment by andrewbranch** [comment]
Decision from design meeting:

1. Let's bring back the commit that changed the declaration emit
2. For back compat reasons, we cannot issue the error on .d.ts files
3. For implementation files, we actually want to issue an error regardless of `noUnusedParameter`/`noUnusedLocals`. We should craft a new error message, not the “declared but never read” message.
4. _However_, we want to silence the error if the rebound name is used in a type query somewhere else in the type:
   ```ts
   type F1 = ({ x: renamedX }) => void;             // Error
   type F2 = ({ x: renamedX }) => typeof renamedX;  // No error
   ```

**Comment by andrewbranch** [comment]
@uhyo are you interested in making these changes, almost 2 years later? 😆

**Comment by uhyo** [comment]
Sure! I'll work on this within a few days.

**Comment by uhyo** [comment]
@andrewbranch I renewed the PR so that it matches the decision from design meeting.

- [Removes unnecessary renaming from declaration emit.](https://github.com/microsoft/TypeScript/pull/41044/files#diff-db472012cfa3a14312104e7706cae299c5ec18bd4978f2c24219f3392d773a5eR74)
- [.d.ts files are not checked.](https://github.com/microsoft/TypeScript/pull/41044/files#diff-db472012cfa3a14312104e7706cae299c5ec18bd4978f2c24219f3392d773a5eR74)
- [Check is done regardless of `noUnusedLocals` or `noUnusedParameters`.](https://github.com/microsoft/TypeScript/pull/41044/files#diff-7ae872724e13d3e320f59d7ca7a1aa9865a4629a117ab94c58b0f471fe891413R1-R2) [Also the error message is new.](https://github.com/microsoft/TypeScript/pull/41044/files#diff-30c30041ec8ab23b1bd6201815a8aa8afba2596ff7ca33217afc318c62ad3a1aR23) (I'm not very good at crafting error messages though 😨)
- [If rebound name is actually used, no error is reported.](https://github.com/microsoft/TypeScript/pull/41044/files#diff-30c30041ec8ab23b1bd6201815a8aa8afba2596ff7ca33217afc318c62ad3a1aR33) [Declaration emit also respects the same rule.](https://github.com/microsoft/TypeScript/pull/41044/files#diff-db472012cfa3a14312104e7706cae299c5ec18bd4978f2c24219f3392d773a5eR141)

I'd like your review 😉

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -3487,6 +3487,10 @@
         "category": "Error",
         "code": 2841
     },
+    "Variable '{0}' is not used. Did you mean to write a type annotation for whole object?": {
```
[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply src/compiler/diagnosticMessages.json:3490]:
```
Variable '{0}' is unused - did you intend to use it as a type annotation for '{1}'?
```

Related span on the end of the object if a type annotation doesn't already exist

```
We can only write a type for '{0}' by adding a type for the destructured parameter here.
```

[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/diagnosticMessages.json:3490]: Oh wait, you can’t get this in a React component, this is only in types 😆

[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/diagnosticMessages.json:3490]: Also, I guess you meant for `{1}` to be just the property name, not the whole object.

`'{0}' is an unused renaming of '{1}'. Did you mean to write a type annotation instead?`

Then with your related span as well? I just wish it was easier to tell where “here” was referring to in the related span.

[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/diagnosticMessages.json:3490]: (Going to slowly iterate on my message until it’s identical to yours and pretend it was my idea)

[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply src/compiler/diagnosticMessages.json:3490]: Maybe a mash-up?

```
'{0}' is an unused renaming of '{1}'. Did you intend to use it as a type annotation?
```

[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/diagnosticMessages.json:3490]: I like this. I also like the related span message. I feel like the use of “here” in related spans can be confusing, but that’s how all of our related spans are, and it’s more of an editor display issue than an issue on our end.

[the remark this answers is not in the document]
Comment (uhyo) [reply src/compiler/diagnosticMessages.json]: I added test cases and updated diagnostic messages!

Does this match your intention?

Screenshot of below snippet


```
src/index.ts:5:17 - error TS2842: 'string' is an unused renaming of 'a'. Did you intend to use it as a type annotation?

5 type F4 = ({ a: string }) => any; // Error
                  ~~~~~~

  src/index.ts:5:24
    5 type F4 = ({ a: string }) => any; // Error
                             ~
    We can only write a type for '{ a: string }' by adding a type for the entire parameter here.
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1026,6 +1026,11 @@ namespace ts {
         const potentialNewTargetCollisions: Node[] = [];
         const potentialWeakMapSetCollisions: Node[] = [];
         const potentialReflectCollisions: Node[] = [];
+        const potentialAlwaysCheckedUnusedTypes: {
```
[the remark this answers is not in the document]
Comment (uhyo) [reply src/compiler/checker.ts:1029]: Thank you for feedback, fixed 🙂

File: tests/cases/compiler/renamingDestructuredPropertyInFunctionType.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+type G4 = new ({ a: string }: O) => any; // Error
+type G5 = new ({ a: string, b, c }: O) => any; // Error
+type G6 = new ({ a: string }) => typeof string; // OK
+type G7 = new ({ a: string, b: number }) => typeof number; // Error
+type G8 = new ({ a, b: number }) => typeof number; // OK
+type G9 = new ([a, b, c]) => void; // Error
+type G10 = new ({ "a": string }) => void; // Error
+type G11 = new ({ 2: string }) => void; // Error
+type G12 = new ({ ["a"]: string }: O) => void; // Error
+type G13 = new ({ [2]: string }) => void; // Error
```
Comment (andrewbranch) [inline tests/cases/compiler/renamingDestructuredPropertyInFunctionType.ts:32]: Can you add a comment explaining that `string` has to be retained in declaration emit even though it’s an error, because eliding it would leave invalid syntax (since the property names aren’t identifiers)?

File: tests/cases/compiler/renamingDestructuredPropertyInFunctionType.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+type G1 = new (arg: number) => any; // OK
+type G2 = new ({ a: string }: O) => any; // Error
+type G3 = new ({ a: string, b, c }: O) => any; // Error
+type G4 = new ({ a: string }: O) => any; // Error
+type G5 = new ({ a: string, b, c }: O) => any; // Error
+type G6 = new ({ a: string }) => typeof string; // OK
+type G7 = new ({ a: string, b: number }) => typeof number; // Error
+type G8 = new ({ a, b: number }) => typeof number; // OK
+type G9 = new ([a, b, c]) => void; // Error
```
Comment (andrewbranch) [inline tests/cases/compiler/renamingDestructuredPropertyInFunctionType.ts:28]: This is incorrectly labeled “Error”

**Review by andrewbranch — CHANGES_REQUESTED** [review]
Almost there 👍

File: src/compiler/diagnosticMessages.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/diagnosticMessages.json]: Does any mark show up for the related span if it’s zero-length? At first I thought the span from `end - 1` to `end` was a little weird, but it shows up pretty well here. I would maybe move it back one character (`[end, end + 1]`).

Also, mentioned this in the review, but I think it should be `We can only write a type for 'a'`.

[the remark this answers is not in the document]
Comment (uhyo) [reply src/compiler/diagnosticMessages.json]: I fixed interpolation and moved the span one step further.

スクリーンショット 2022-06-10 11 10 14

I also tested zero-length span but it didn't show any mark.

**Review by andrewbranch — APPROVED** [review]
Thank you! 🌟

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by Mister-Hope** [comment]
I think https://github.com/microsoft/TypeScript/issues/50483 is probably cause by this PR, could anyone check it?
